### PR TITLE
feat: support login

### LINKED ISSUE #31: Support login
Design discussion: #18 

Proposed API:
```go
package credentials

func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
	panic("not implemented") // TODO: Implement
}
```

References:

- [Notation login](https://github.com/notaryproject/notation/blob/main/cmd/notation/login.go#L68-L84)
- [Oras login](https://github.com/oras-project/oras/blob/main/cmd/oras/root/login.go#L111-L129)

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Resolves #31

### CODE REVIEW

File: login.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	if err := registry.Ping(ctx); err != nil {
+		return fmt.Errorf("unable to connect to the registry: %v", err)
+	}
+	if err := store.Put(ctx, registry.Reference.Registry, cred); err != nil {
+		return fmt.Errorf("unable to store the credential: %v", err)
+	}
+	fmt.Println("Login Succeeded")
```
Comment (shizhMSFT) [inline login.go:34]: library should never print stuffs.

Comment (wangxiaoxuan273) [reply login.go:34]: removed

File: login.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	"fmt"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	if err := registry.Ping(ctx); err != nil {
+		return fmt.Errorf("unable to connect to the registry: %v", err)
```
Comment (Wwwsylvia) [inline login.go:29]: We may want to wrap the error using `%w`. See https://pkg.go.dev/fmt#Errorf

```suggestion
		return fmt.Errorf("unable to connect to the registry: %w", err)
```

File: login.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"context"
+	"fmt"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
```
Comment (Wwwsylvia) [inline login.go:27]: Wondering if we should have a file called `registry.go` and put methods like `Login`, `Logout` and `Credential` in it?

File: login_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func TestLogin(t *testing.T) {
+	// create a test registry
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	defer ts.Close()
+	uri, _ := url.Parse(ts.URL)
+	reg, err := remote.NewRegistry(uri.Host)
+	if err != nil {
+		panic("cannot create test registry")
```
Comment (Wwwsylvia) [inline login_test.go:51]:
```suggestion
		t.Fatalf("cannot create test registry: %v", err)

```

File: login_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			registry: *reg,
+			cred:     auth.Credential{Username: testUsername, Password: testPassword},
+			wantErr:  true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := Login(tt.ctx, tt.store, tt.registry, tt.cred); (err != nil) != tt.wantErr {
+				t.Errorf("Login() error = %v, wantErr %v", err, tt.wantErr)
+			}
```
Comment (Wwwsylvia) [inline login_test.go:86]: We can have `testStore` actually store the credential, and check if the credential is stored after `Login()`.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: login.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	if err := registry.Ping(ctx); err != nil {
+		return fmt.Errorf("unable to connect to the registry: %v", err)
+	}
+	if err := store.Put(ctx, registry.Reference.Registry, cred); err != nil {
```
Comment (Wwwsylvia) [inline login.go:31]: Should we support redirecting `docker.io` like oras does? See
https://github.com/oras-project/oras/blob/main/cmd/oras/root/login.go#L120-L129 /cc @qweeah

File: login.go (hunk 3 above)
Comment (wangxiaoxuan273) [reply login.go:27]: Created this file.

File: login_test.go (hunk 5 above)
Comment (wangxiaoxuan273) [reply login_test.go]: Added the store functionality.

File: registry.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	if err := registry.Ping(ctx); err != nil {
+		return fmt.Errorf("unable to login to the registry %s: %w", registry.Reference.Registry, err)
+	}
+	hostname := registry.Reference.Registry
+	if hostname == "docker.io" {
+		hostname = "https://index.docker.io/v1/"
+	}
+	if err := store.Put(ctx, hostname, cred); err != nil {
```
Comment (Wwwsylvia) [inline registry.go:35]: nit: We may need some comments for this part.

File: registry_test.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := Login(tt.ctx, ns, tt.registry, tt.cred); (err != nil) != tt.wantErr {
+				t.Fatalf("Login() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !reflect.DeepEqual(ns.storage[tt.registry.Reference.Registry], tt.cred) {
+				t.Fatalf("Incorrect credentials stored")
+			}
```
Comment (Wwwsylvia) [inline registry_test.go:95]: nit: We should print both the actual value and the expected value, something like
```suggestion
			if got := ns.storage[tt.registry.Reference.Registry]; !reflect.DeepEqual(got, tt.cred) {
				t.Errorf("Stored credential = %v, want %v", got, tt.cred)
			}
```

File: login.go (hunk 6 above)
Comment (wangxiaoxuan273) [reply login.go]: Added redirect for docker.io

File: registry.go (hunk 7 above)
Comment (wangxiaoxuan273) [reply registry.go:35]: Added the comments

File: registry.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	if err := registry.Ping(ctx); err != nil {
+		return fmt.Errorf("unable to login to the registry %s: %w", registry.Reference.Registry, err)
+	}
+	hostname := mapHostname(registry.Reference.Registry)
+	if err := store.Put(ctx, hostname, cred); err != nil {
+		return fmt.Errorf("unable to store the credential: %v", err)
```
Comment (Wwwsylvia) [inline registry.go:33]: nit: How about
```suggestion
		return fmt.Errorf("unable to store the credential for %s: %w", hostname, err)
```

File: registry_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+			cred:     auth.Credential{Username: testUsername, Password: testPassword},
+			wantErr:  true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := Login(tt.ctx, ns, tt.registry, tt.cred); (err != nil) != tt.wantErr {
+				t.Fatalf("Login() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if got := ns.storage[tt.registry.Reference.Registry]; !reflect.DeepEqual(got, tt.cred) {
```
Comment (Wwwsylvia) [inline registry_test.go:93]: I just realized that the same registry is used for both the success test and the failure test, so this check will always be true. 🤔

Comment (wangxiaoxuan273) [reply registry_test.go:93]: Good point! Thanks for spotting it!

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	"context"
+	"fmt"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	if err := registry.Ping(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply registry.go:34]: The word `registry` is a bit overloaded here although I don't have a better suggestion.

File: registry.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	if err := store.Put(ctx, hostname, cred); err != nil {
+		return fmt.Errorf("unable to store the credential for %s: %w", hostname, err)
+	}
+	return nil
+}
+
+func mapHostname(hostname string) string {
+	// For a user case that login 'docker.io',
+	// According the the behavior of Docker CLI,
+	// credential should be added under key "https://index.docker.io/v1/"
```
Comment (TerryHowe) [inline registry.go:47]:
```suggestion
	// The Docker CLI expects that the 'docker.io' credential
	// will be added under the key "https://index.docker.io/v1/"
```

**Review by Wwwsylvia — APPROVED** [review]
LGTM

File: registry.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry.go:34]: > The word `registry` is a bit overloaded here although I don't have a better suggestion.

`reg` might be good enough

[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry.go:34]: changed to `reg`

File: registry.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	"oras.land/oras-go/v2/registry/remote/retry"
+)
+
+// Login provides the login functionality with the given credentials.
+func Login(ctx context.Context, store Store, registry remote.Registry, cred auth.Credential) error {
+	registry.Client = &auth.Client{
+		Client:     retry.DefaultClient,
+		Cache:      auth.DefaultCache,
+		Credential: auth.StaticCredential(registry.Reference.Registry, cred),
+	}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry.go]: Used a copied client local to the `login` function.

File: registry.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	"context"
+	"errors"
+	"fmt"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// Login provides the login functionality with the given credentials. A local client is used
+// to keep the original client of the registry intact.
```
Comment (shizhMSFT) [inline registry.go:28]: "A local client is used to keep the original client of the registry intact." requires rewording.

BTW, we require the `reg.Client` to be nil or of `*auth.Client` type. Basically, developers will not read your code but read your documentation before invoking the API.

File: registry.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+		return fmt.Errorf("unable to login to the registry %s: %w", regClone.Reference.Registry, err)
+	}
+	hostname := mapHostname(regClone.Reference.Registry)
+	if err := store.Put(ctx, hostname, cred); err != nil {
+		return fmt.Errorf("unable to store the credential for %s: %w", hostname, err)
+	}
+	return nil
+}
+
+func mapHostname(hostname string) string {
```
Comment (shizhMSFT) [inline registry.go:56]: note: let's keep `mapHostname()` here for now. You will need another `mapHostname` later.

File: registry.go (hunk 14 above)
Comment (wangxiaoxuan273) [reply registry.go:28]: changed the comment.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**